src/core/net/dns_client.cpp

```diff
@@ -815,20 +860,60 @@ void Client::SendQuery(Query &aQuery, QueryInfo &aInfo, bool aUpdateTimer)
         SuccessOrExit(error = message->Append(Question(kQuestionRecordTypes[aInfo.mQueryType][num])));
     }
 
-    messageInfo.SetPeerAddr(aInfo.mConfig.GetServerSockAddr().GetAddress());
-    messageInfo.SetPeerPort(aInfo.mConfig.GetServerSockAddr().GetPort());
+    length = message->GetLength() - message->GetOffset();
+#if OPENTHREAD_CONFIG_DNS_OVER_TCP_ENABLE
 
-    SuccessOrExit(error = mSocket.SendTo(*message, messageInfo));
+    // Send udp query in case of message being greater than kUdpQueryMaxSize or as last resort.
+    if (length >= kUdpQueryMaxSize || aInfo.mTransmissionCount >= aInfo.mConfig.GetMaxTxAttempts() ||
+        aInfo.mConfig.GetTransportProto() == QueryConfig::kDnsTcpTransport)
+    {
+        mSendLink.mNext = nullptr;
+        mSendLink.mData = mSendBufferBytes;
```

Comment:
@abtink Thanks for reviewing. 
I think that this is the concern you had [here](https://github.com/openthread/openthread/pull/8474#pullrequestreview-1221937567)

All in all, we can add multiple endpoints to this implementation but as I understand for each of those we would have to preallocate manually RX and TX buffers. However, this introduces unwanted memory consumption. 

One of the solutions would be to refactor the TCP stack to utilize `otMessage`.
Alternatively, we can limit the number of servers that we are connecting to at the same time, rather than the queries that we are issuing. 

I would be in favor of the second option.